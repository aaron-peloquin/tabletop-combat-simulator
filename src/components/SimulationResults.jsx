import React, {Fragment} from "react"
import {connect} from "react-redux"
import {
  Grid,
  Typography,
} from "@material-ui/core"
import {withStyles} from "@material-ui/core/styles"

const Styles = (theme) => {
  return {
    CreatureName: {
      textAlign: "left",
    },
    CreatureHealth: {
      textAlign: "right",
      paddingRight: theme.spacing.unit,
    },
  }
}

const SimulationResults = (props) => {
  const {classes, Victory, Survivors} = props

  return <Fragment>
    {(Victory.length > 0?<Typography variant="h4">Team <strong>{Victory.toUpperCase()}</strong> Won</Typography>:"")}
    {(Survivors.length > 0?<Grid container>
      {Survivors.map((Creature, k) => {
        return <Fragment key={k}>
          <Grid className={classes.CreatureHealth} item xs={6}>{Creature.Health} HP</Grid>
          <Grid className={classes.CreatureName} item xs={6}>{Creature.Name}</Grid>
        </Fragment>
      })}</Grid>:"")}
  </Fragment>
}

const MapStateToProps = (state) => {
  const Victory = state.combat.Victory
  let Survivors = false
  if (Victory.length > 0) {
    Survivors = state.combat.AliveTeamCreatures[Victory].map((hash) => {
      const creature = state.combat.CreatureStatus[hash]
      return {
        Name: creature.name,
        Health: creature.hp,
      }
    })
  }

  return {
    Victory,
    Survivors,
  }
}

const ConnectedComponent = connect(MapStateToProps)(SimulationResults)
const StyledComponent = withStyles(Styles)(ConnectedComponent)

export default StyledComponent
