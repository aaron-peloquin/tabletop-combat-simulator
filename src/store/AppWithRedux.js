import React from 'react'
import {initializeStore} from './store'

const isServer = typeof window === 'undefined'
const reduxStoreSymbol = Symbol("reduxStore")

function getOrCreateStore (initialState) {
  // Always make a new store if server, otherwise state is shared between requests
  if (isServer) {
    return initializeStore(initialState)
  }

  // Create store if unavailable on the client and set it on the window object
  if (!window[reduxStoreSymbol]) {
    window[reduxStoreSymbol] = initializeStore(initialState)
  }
  return window[reduxStoreSymbol]
}

export default (App) => {
  return class AppWithRedux extends React.Component {
    static async getInitialProps (appContext) {
      // Get or Create the store with `undefined` as initialState
      // This allows you to set a custom default initialState
      const reduxStore = getOrCreateStore()

      // Provide the store to getInitialProps of pages
      appContext.ctx.reduxStore = reduxStore

      let appProps = {}
      if (typeof App.getInitialProps === 'function') {
        appProps = await App.getInitialProps(appContext)
      }

      return {
        ...appProps,
        initialReduxState: reduxStore.getState()
      }
    }

    constructor (props) {
      super(props)
      this.reduxStore = getOrCreateStore(props.initialReduxState)
    }

    render () {
      return <App {...this.props} reduxStore={this.reduxStore} />
    }
  }
}
