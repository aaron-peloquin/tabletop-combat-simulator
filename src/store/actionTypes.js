/**
 * This is an object of Redux Actions. It mostly exists to
 *  ensure we do not create duplicate actions, but it can
 *  also be used as a list of our existing actions
*/
const actions = {
  CopyCreature: 'CopyCreature',
  DeleteAllCreatures: 'DeleteAllCreatures',
  DeleteCreature: 'DeleteCreature',
  SaveCreature: 'SaveCreature',
  SetEditCreature: 'SetEditCreature',
  ToggleSidebar: 'ToggleSidebar',
  RunSimulation: 'RunSimulation',
  UpdateEditCreature: 'UpdateEditCreature',
}

export default actions
