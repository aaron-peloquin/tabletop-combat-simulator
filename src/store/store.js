import {createStore, applyMiddleware, combineReducers} from "redux"
import {composeWithDevTools} from "redux-devtools-extension"
import thunkMiddleware from "redux-thunk"

/** Reducers */
import combat from "./reducer/combat"
import creatures from "./reducer/creatures"
import sideBar from "./reducer/sideBar"
import editing from "./reducer/editing"

/** Reducers */
const reducers = combineReducers({
  combat,
  creatures,
  editing,
  sideBar,
})

/**
 * Initializes the store
 * @param {obj} initialState the state our app starts with
 * @return {obj} the redux store
 */
export function initializeStore(initialState = initialState) {
  return createStore(reducers, initialState, composeWithDevTools(applyMiddleware(thunkMiddleware)))
}
