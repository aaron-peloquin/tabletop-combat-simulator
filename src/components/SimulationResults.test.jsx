import {Provider} from "react-redux"
import {shallow} from "enzyme"
import toJson from "enzyme-to-json"
import {initializeStore} from "../store/store"

import {SimulationResults, MapStateToProps} from "./SimulationResults"

describe("<SimulationResults />", () => {
  let store; let SimulationResultsComponent

  beforeEach(() => {
    store = initializeStore({})

    SimulationResultsComponent = shallow(<SimulationResults
      classes={{
        CreatureHealth: `color: "blue"`,
        CreatureName: `color: "orange"`,
      }}
      Victory="A"
      FinalRound={4}
      Survivors={[
        {Name: "Beth", Health: 25},
        {Name: "Jerry", Health: 50},
        {Name: "Frank", Health: 75},
      ]}
    />)
  })

  it("renders", () => {
    expect(typeof SimulationResultsComponent).toBe("object")
  })

  it("snapshots", () => {
    const Tree = toJson(SimulationResultsComponent)
    expect(Tree).toMatchSnapshot()
  })

  describe('MapStateToProps()', () => {
    it('returns correctly', () => {
      const mockState = {
        combat: {
          FinalRound: 5,
          Victory: "D",
          CreatureStatus: {
            somehash: {
              name: "Goplin",
              hp: 4,
            }
          },
          AliveTeamCreatures: {"D": ["somehash"]},
        },
      }
      const result = MapStateToProps(mockState)
      })
  })
})
