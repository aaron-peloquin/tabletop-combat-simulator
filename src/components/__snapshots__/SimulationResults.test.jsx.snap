// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SimulationResults /> snapshots 1`] = `
<Fragment>
  <WithStyles(Typography)
    variant="h4"
  >
    Team 
    <strong>
      A
    </strong>
     Won
  </WithStyles(Typography)>
  <WithStyles(Typography)
    variant="body1"
  >
    Combat resolved after 
    <strong>
      4
    </strong>
     rounds.
  </WithStyles(Typography)>
  <WithStyles(Grid)
    container={true}
  >
    <WithStyles(Grid)
      className="color: \\"blue\\""
      item={true}
      xs={6}
    >
      25
       HP
    </WithStyles(Grid)>
    <WithStyles(Grid)
      className="color: \\"orange\\""
      item={true}
      xs={6}
    >
      Beth
    </WithStyles(Grid)>
    <WithStyles(Grid)
      className="color: \\"blue\\""
      item={true}
      xs={6}
    >
      50
       HP
    </WithStyles(Grid)>
    <WithStyles(Grid)
      className="color: \\"orange\\""
      item={true}
      xs={6}
    >
      Jerry
    </WithStyles(Grid)>
    <WithStyles(Grid)
      className="color: \\"blue\\""
      item={true}
      xs={6}
    >
      75
       HP
    </WithStyles(Grid)>
    <WithStyles(Grid)
      className="color: \\"orange\\""
      item={true}
      xs={6}
    >
      Frank
    </WithStyles(Grid)>
  </WithStyles(Grid)>
</Fragment>
`;
